checker/src/main/java/org/checkerframework/checker/nullness/NullnessTransfer.java

```diff
@@ -357,21 +358,35 @@ public TransferResult<NullnessValue, NullnessStore> visitMethodInvocation(
             MethodInvocationNode n, TransferInput<NullnessValue, NullnessStore> in) {
         TransferResult<NullnessValue, NullnessStore> result = super.visitMethodInvocation(n, in);
 
-        // Make receiver non-null.
+        MethodInvocationTree tree = n.getTree();
+        ExecutableElement method = TreeUtils.elementFromUse(tree);
+        boolean nonNullAssumption =
+                !analysis.getTypeFactory()
+                        .getChecker()
+                        .hasOption("receiverAndParamsNullableAfterMethodCall");
+
+        boolean isMethodSideEffectFree = PurityUtils.isSideEffectFree(atypeFactory, method);
         Node receiver = n.getTarget().getReceiver();
-        makeNonNull(result, receiver);
+        if (nonNullAssumption
+                || JavaExpression.fromNode(receiver).isUnassignableByOtherCode()
```

Comment:
Switch this check and the next one: checking the boolean `isMethodSideEffectFree` is faster than building the JavaExpression.